### PR TITLE
Refactoring on connection pool

### PR DESCRIPTION
**Pull request by zhenlineo**
Made conneciton pool lock free
Added ConcurrentSet impl based on ConcurrentDict
Improved the impl of RRSet to use list to avoid memory copy

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/Internal/ConcurrentSet.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace Neo4j.Driver.Internal
+{
+    internal class ConcurrentSet <T> : IEnumerable<T>
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConcurrentSet.cs:7]: Should `ConcurrentSet` maybe implement `ISet`?

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 2]
```diff
@@ -223,24 +217,20 @@ protected override void Dispose(bool isDisposing)
...
+                foreach (var inUseConnection in _inUseConnections)
                 {
-                    var connections = new List<IPooledConnection>(_inUseConnections.Values);
-                    _inUseConnections.Clear();
-                    foreach (var inUseConnection in connections)
+                    Logger?.Info($"Disposing In Use Connection {inUseConnection.Id}");
+                    if (_inUseConnections.TryRemove(inUseConnection))
                     {
-                        Logger?.Info($"Disposing In Use Connection {inUseConnection.Id}");
                         inUseConnection.Close();
```
Comment (lutovich) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs:225]: Can this call throw an exception?

**Review by lutovich — APPROVED** [review]
@zhenlineo changes look good. Made couple small questions.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConcurrentSet.cs (hunk 1 above)
Comment (zhenlineo) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConcurrentSet.cs:7]: We could, then we need to impl 20+ methods on ISet interface. I feel it is not needed.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs (hunk 2 above)
Comment (zhenlineo) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs:225]: In theory, no, as all impl of `IConnection.Close` method will only log errors but not throw any out.

**Pull request merged**